temporal-kotlin/README.md

```diff
@@ -17,4 +19,84 @@ Add temporal-kotlin as a dependency to your pom.xml:
 or to build.gradle:
 ```
 compile group: 'io.temporal', name: 'temporal-kotlin', version: 'N.N.N'
-```
\ No newline at end of file
+```
+
+## Kotlin extensions
+
+This module adds several Kotlin extensions to make Kotlin code that uses Temporal Java SDK a bit
+more idiomatic.
+
+### Options classes DSL
+
+Various Options classes from the SDK can be fluently instantiated with a constructor-like DSL.
+There's also a `copy` extension method that creates a new copy of an Options instance with some
+overridden attributes. Overall, the DSL should make Options classes feel somewhat like Kotlin data
+classes.
```

Comment:
> Couldn't you have something like
> 
> ```kotlin
> data class WorkflowOptionData(val workflowId: String, val taskQueue: String? = null) {
>   toOptions(): WorkflowOption { /* ... */ }
> }
> 
> val options = WorkflowOptionData(
>   workflowId: "ID",
>   taskQueue: if (shouldOverrideTaskQueue) "newQueue" else null,
> )
> ```

It could be, depending on the specific options class and the specific property, but in general, setting a property to `null` is not always equivalent to not setting it at all.

I played a bit with the idea of having proper data classes for options, but so far don't see a strong justification for it. The most significant downside that I see is that adding new fields to data classes [is not a binary backwards-compatible operation](https://jakewharton.com/public-api-challenges-in-kotlin/).

Reply:
> I am unsure if we expect binary compatibility vs recompile when using Java, but if it is a concern, data classes can be avoided. Arguably a more Kotlin-like builder/constructor for each object w/ defaults would still have ergonomic value over calling setters, but I don't have that strong of an opinion on it.

We could do that with the existing approach, but it would require a bit of boilerplate:
```kotlin
var WorkflowOptions.Builder.workflowId: String
  get() { throw UnsupportedOperationException("Reading properties inside builder is not supported") }
  set(value) {
    setWorkflowId(value)
  }
```

The good news is that could be added in a backward-compatible way, so if there's a request for terser/more natural DSL, we could add that later on.